### PR TITLE
fix: upscale opacity

### PR DESCRIPTION
**Pull request by cl-shaharshavit-rnd**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔                                                                        |
| New feature     | ✖                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

Upscale fix
Origin fix in some methods
Convert merged tiles to PNG with Alpha
Fix relevant tests

### CODE REVIEW

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 1]
```diff
@@ -9,25 +9,52 @@ public class TileScaler : ITileScaler
...
+            //find source pixels range
+            double tilePartX = targetCoords.X % scale;
+            double tilePartY = targetCoords.Y % scale;
+            double tileSize = TILE_HEIGHT / (double)scale;
 
-            int pixleX = (int)(tilePartX * TILE_WIDTH);
-            int pixleY = (int)(tilePartY * TILE_HEIGHT);
-            int imageWidth = Math.Max((int)tileSize, 1);
-            int imageHeight = Math.Max((int)tileSize, 1);
+            int pixelX = (int)(tilePartX * tileSize);
```
Comment (asafMasa) [inline MergerLogic/ImageProcessing/TileScaler.cs:22]: Why did you remove the TILE_WIDTH? is it valid in all conditions? what if TILE_HEIGHT won't be equal to TILE_WIDTH?

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 2]
```diff
@@ -9,25 +9,52 @@ public class TileScaler : ITileScaler
...
+            int pixelY = (int)(tilePartY * tileSize);
```
Comment (asafMasa) [inline MergerLogic/ImageProcessing/TileScaler.cs:23]: why did the calculation changed from 'TILE_HEIGHT' to 'tileSize'?

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 3]
```diff
@@ -9,25 +9,52 @@ public class TileScaler : ITileScaler
         private const int TILE_WIDTH = 256;
         private const int TILE_HEIGHT = 256;
 
-        public void Upscale(MagickImage baseImage, Tile baseTile, Coord targetCoords)
+        public MagickImage Upscale(MagickImage baseImage, Tile baseTile, Coord targetCoords)
```
Comment (asafMasa) [inline MergerLogic/ImageProcessing/TileScaler.cs:12]: add comment that includes description and link to chosen upscale algorithm

File: MergerLogic/ImageProcessing/TileMerger.cs [hunk 4]
```diff
@@ -158,7 +164,9 @@ private void AddTileToImageList(Coord targetCoords, Tile? tile, List<MagickImage
             tileImage = new MagickImage(tileBytes);
             if (tile.Z < targetCoords.Z)
             {
-                this._tileScaler.Upscale(tileImage, tile, targetCoords);
+                var upscale = this._tileScaler.Upscale(tileImage, tile, targetCoords);
```
Comment (asafMasa) [inline MergerLogic/ImageProcessing/TileMerger.cs:167]: why don't you dispose and replace the "tileImage" inside the method? every one that uses this method needs to remember to dispose it himself - can lead to bugs

**Review by asafMasa — CHANGES_REQUESTED** [review]
👍 anxious to see the results

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 1 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProcessing/TileScaler.cs:22]: are you aware of any tile grid that have none square tiles?

File: MergerLogic/ImageProcessing/TileMerger.cs (hunk 4 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProcessing/TileMerger.cs:167]: 1. replacing inside the function requires making the pram ref.
2. what if the calling function still want to use the original tile as well as the upscaled? disposing should be handler in the same level as the allocation

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 3 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProcessing/TileScaler.cs:12]: the "algorithm" is just to enlarge the pixels...  its soo simple and basic calling it "algorithm" is overstatement and its trivial enough that I doubt the will be dedicated page anywhere (and linking to page with multiple algorithms is more confusing then helpful in this context

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 2 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProcessing/TileScaler.cs:23]: because tiles are square

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 5]
```diff
@@ -9,25 +9,52 @@ public class TileScaler : ITileScaler
...
         {
             int zoomLevelDiff = targetCoords.Z - baseTile.Z;
             int scale = 1 << zoomLevelDiff;
-            double scaleAsDouble = (double)scale;
 
-            double tilePartX = (targetCoords.X % scale) / scaleAsDouble;
-            double tilePartY = (targetCoords.Y % scale) / scaleAsDouble;
-            double tileSize = TILE_HEIGHT / scaleAsDouble;
+            //find source pixels range
+            double tilePartX = targetCoords.X % scale;
```
Comment (shimoncohen) [inline MergerLogic/ImageProcessing/TileScaler.cs:18]: Why did you remove the part dividing by the scale?
The scale is not absolute, it is the difference in scales, so it should be divided by.

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 6]
```diff
@@ -9,25 +9,52 @@ public class TileScaler : ITileScaler
         private const int TILE_WIDTH = 256;
```
Comment (shimoncohen) [inline MergerLogic/ImageProcessing/TileScaler.cs:9]: Replace with `TILE_SIZE = 256`.

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 1 above)
Comment (shimoncohen) [inline MergerLogic/ImageProcessing/TileScaler.cs:22]: This is incorrect because of my last comment, it will not give you a value in the correct bounds.

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 7]
```diff
@@ -9,25 +9,52 @@ public class TileScaler : ITileScaler
...
+            double tileSize = TILE_HEIGHT / (double)scale;
 
-            int pixleX = (int)(tilePartX * TILE_WIDTH);
-            int pixleY = (int)(tilePartY * TILE_HEIGHT);
-            int imageWidth = Math.Max((int)tileSize, 1);
-            int imageHeight = Math.Max((int)tileSize, 1);
+            int pixelX = (int)(tilePartX * tileSize);
+            int pixelY = (int)(tilePartY * tileSize);
+            int srcSize = Math.Max((int)tileSize, 1);
+            int maxSrcX = pixelX + srcSize;
```
Comment (shimoncohen) [inline MergerLogic/ImageProcessing/TileScaler.cs:25]: Here too, you are out of bounds of 256.

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 8]
```diff
@@ -9,25 +9,52 @@ public class TileScaler : ITileScaler
...
-            MagickGeometry geometry = new MagickGeometry(pixleX, pixleY, imageWidth, imageHeight);
-            baseImage.Crop(geometry);
-            baseImage.RePage();
-            baseImage.Resize(TILE_WIDTH, TILE_HEIGHT);
+            //prepare pixels data 
+            var scaledImage = new MagickImage(MagickColor.FromRgba(0, 0, 0, 0), TILE_WIDTH, TILE_HEIGHT);
+            scaledImage.HasAlpha = baseImage.HasAlpha;
+            var srcPixels = baseImage.GetPixels();
+            var targetPixels = scaledImage.GetPixels();
+            int byteCount = 4 * scale * scale;
```
Comment (shimoncohen) [inline MergerLogic/ImageProcessing/TileScaler.cs:33]: 4 times the square of the zoom level scale difference? why?

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 5 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProcessing/TileScaler.cs:18]: it want removed it was optimized.
the old calculation divided it by scale and then multiplied it by the base tile size(256)
now it is multiplied by the sub tile size instead (256/scale) so the calculation wasn't changed but less operations are needed

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 1 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProcessing/TileScaler.cs:22]: it is not because of my answer to your last comment

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 8 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProcessing/TileScaler.cs:33]: 4 color channels (RGBA), with 1 byte each, and both x and y axis are multiplied by the scale so it is squared

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 9]
```diff
@@ -6,28 +6,76 @@ namespace MergerLogic.ImageProcessing
...
+            double tilePartY = scale - 1 - (targetCoords.Y % scale);// flip direction as tiles are LL and pixels are UL
+            double subTileSize = TILE_SIZE / (double)scale;
 
-            int pixleX = (int)(tilePartX * TILE_WIDTH);
-            int pixleY = (int)(tilePartY * TILE_HEIGHT);
-            int imageWidth = Math.Max((int)tileSize, 1);
-            int imageHeight = Math.Max((int)tileSize, 1);
+            int pixelX = (int)(tilePartX * subTileSize);
+            int pixelY = (int)(tilePartY * subTileSize);
+            int srcSize = Math.Max((int)subTileSize, 1);
```
Comment (asafMasa) [inline MergerLogic/ImageProcessing/TileScaler.cs:23]: remove max

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 10]
```diff
@@ -6,28 +6,76 @@ namespace MergerLogic.ImageProcessing
...
-            baseImage.Crop(geometry);
-            baseImage.RePage();
-            baseImage.Resize(TILE_WIDTH, TILE_HEIGHT);
+            var srcPixels = baseImage.GetPixels();
+            MagickImage scaledImage;
+            var colorSpace = baseImage.ColorSpace;
+            var colorType = baseImage.ColorType;
+            baseImage.ColorType = baseImage.HasAlpha ? ColorType.TrueColorAlpha : ColorType.TrueColor; 
+            baseImage.ColorSpace = ColorSpace.sRGB;
+            if (scale >= 256)
```
Comment (asafMasa) [inline MergerLogic/ImageProcessing/TileScaler.cs:31]: replace with TILE_SIZE

**Review by asafMasa — CHANGES_REQUESTED** [review]

**Review by shimoncohen — APPROVED** [review]

**Review by asafMasa — APPROVED** [review]
